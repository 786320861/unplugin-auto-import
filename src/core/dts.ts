import { ImportsFlatMap } from '../types'

export function generateDeclration(imports: ImportsFlatMap) {
  const body = Object
    .entries(imports)
    .map(([name, info]) => `  const ${name}: typeof import('${info.module}')['${info.from || name}']`)
    .join('\n')
  return `// Generated by 'unplugin-auto-import'\ndeclare global {\n${body}\n}\nexport {}\n`
}
