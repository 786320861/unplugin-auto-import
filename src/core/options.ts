import { resolve } from 'path'
import { toArray } from '@antfu/utils'
import { createFilter } from '@rollup/pluginutils'
import { presets } from '../presets'
import { ImportInfo, ImportsFlatMap, Options, ResolvedOptions } from '../types'

export function resolveOptions(options: Options = {}): ResolvedOptions {
  const imports = flattenImportsMap(options.imports)
  const resolved: ResolvedOptions = {
    ...options,
    dts: options.dts === false
      ? false
      : resolve(options.dts || 'auto-imports.d.ts'),
    imports,
    matchRE: new RegExp(`\\b(${Object.keys(imports).join('|')})\\b`, 'g'),
    idFilter: createFilter(
      options.include || [/\.[jt]sx?$/, /\.vue\??/],
      options.exclude || [/node_modules/, /\.git/],
    ),
  }

  return resolved
}

export function flattenImportsMap(map: Options['imports']): ImportsFlatMap {
  const flat: ImportsFlatMap = {}
  toArray(map).forEach((definition) => {
    if (typeof definition === 'string') {
      if (!presets[definition])
        throw new Error(`[global-import] preset ${definition} not found`)
      definition = presets[definition]
    }

    for (const mod of Object.keys(definition)) {
      for (const id of definition[mod]) {
        const meta = {
          module: mod,
        } as ImportInfo
        if (Array.isArray(id)) {
          meta.name = id[1]
          meta.from = id[0]
        }
        else {
          meta.name = id
        }

        if (flat[meta.name])
          throw new Error(`[global-import] identifier ${meta.name} already defined with ${flat[meta.name].module}`)

        flat[meta.name] = meta
      }
    }
  })

  return flat
}
