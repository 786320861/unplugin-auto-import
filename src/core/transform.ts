import MagicString from 'magic-string'
import { ImportInfo, TransformOptions } from '../types'

const excludeRegex = [
  // imported from other module
  /\bimport\s*\{([\s\S]*?)\}\s*from\b/g,
  // defined as function
  /\bfunction\s*([\s\S]+?)\s*\(/g,
  // defined as local variable
  /\b(?:const|let|var)\s*([\w\d_$]+?)\b/g,
]

export function transform(code: string, id: string, { matchRE, imports }: TransformOptions) {
  const matched = new Set(Array.from(code.matchAll(matchRE)).map(i => i[1]))

  // remove those already defined
  for (const regex of excludeRegex) {
    Array.from(code.matchAll(regex))
      .flatMap(i => i[1]?.split(',') || [])
      .forEach(i => matched.delete(i.trim()))
  }

  const modules: Record<string, ImportInfo[]> = {}

  // group by module name
  Array.from(matched).forEach((name) => {
    const moduleName = imports[name]?.module
    if (!moduleName)
      return
    if (!modules[moduleName])
      modules[moduleName] = []
    modules[moduleName].push(imports[name])
  })

  // stringify import
  const importStatements = Object.entries(modules)
    .map(([moduleName, names]) => {
      const imports = names
        .map(({ name, from }) => from ? `${from} as ${name}` : name)
        .join(', ')
      return `import { ${imports} } from '${moduleName}';`
    })
    .join('')

  const s = new MagicString(code)
  s.prependLeft(0, importStatements)

  return {
    code: s.toString(),
    get map() {
      return s.generateMap()
    },
  }
}
