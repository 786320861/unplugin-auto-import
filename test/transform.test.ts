import { resolve } from 'path'
import { promises as fs } from 'fs'
import fg from 'fast-glob'
import { transform } from '../src/core/transform'
import { resolveOptions } from '../src/core/options'

describe('transform', () => {
  const options = resolveOptions({
    imports: [
      'vue',
      'react',
      'svelte',
      'svelte/animate',
      'svelte/easing',
      'svelte/motion',
      'svelte/store',
      'svelte/transition',
      {
        custom: [
          'customNamed',
          ['default', 'customDefault'],
        ],
      },
    ],
  })

  describe('fixtures', () => {
    const root = resolve(__dirname, 'fixtures')
    const files = fg.sync('*', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      it(file, async() => {
        const fixture = await fs.readFile(resolve(root, file), 'utf-8')
        const pass1 = transform(fixture, file, options)?.code ?? fixture
        expect(pass1).toMatchSnapshot()
        const pass2 = transform(pass1, file, options)?.code ?? pass1
        expect(pass2).toBe(pass1)
      })
    }
  })
})
