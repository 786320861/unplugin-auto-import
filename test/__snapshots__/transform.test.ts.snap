// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform fixtures custom.js 1`] = `
"import { default as customDefault, customNamed } from 'custom';const a = customDefault(customNamed())
"
`;

exports[`transform fixtures react.jsx 1`] = `
"import { useState } from 'react';export function Component() {
  const [count, setCount] = useState()

  return <div>{ count }</div>
}
"
`;

exports[`transform fixtures vue.js 1`] = `
"import { ref } from 'vue';const a = ref(0)
"
`;
